### PR TITLE
#1323: new redis modules

### PR DESCRIPTION
**Pull request by uweschaefer**

### CODE REVIEW

File: factcast-test/src/main/java/org/factcast/test/FactCastExtension.java [hunk 1]
```diff
@@ -15,61 +15,88 @@
...
+    for (FactCastIntegrationTestExtension e : extensions) {
+      e.beforeAll(context);
+    }
   }
 
-  private static boolean isPostgresContainer(Field f) {
-    return ModifierSupport.isStatic(f) && PostgreSQLContainer.class.isAssignableFrom(f.getType());
+  private void initialize() {
+    val discovered =
+        Lists.newArrayList(ServiceLoader.load(FactCastIntegrationTestExtension.class).iterator());
```
Comment (samba2) [inline factcast-test/src/main/java/org/factcast/test/FactCastExtension.java:77]: interesting

File: factcast-factus-redis/src/main/java/org/factcast/factus/redis/UUIDCodec.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import java.nio.ByteBuffer;
+import java.util.UUID;
+import lombok.val;
+import org.redisson.client.codec.BaseCodec;
+import org.redisson.client.protocol.Decoder;
+import org.redisson.client.protocol.Encoder;
+
+public class UUIDCodec extends BaseCodec {
+
+  public static final UUIDCodec INSTANCE = new UUIDCodec();
```
[the remark this answers is not in the document]
Comment (uweschaefer) [reply factcast-factus-redis/src/main/java/org/factcast/factus/redis/UUIDCodec.java:16]: you can/should use it in prod code. Providing (stateful) codecs as singleton is following the redisson pattern that suggests this

File: factcast-test/src/main/java/org/factcast/test/FactCastExtension.java [hunk 3]
```diff
@@ -15,61 +15,88 @@
...
 import org.junit.jupiter.api.extension.*;
-import org.junit.platform.commons.support.ModifierSupport;
-import org.junit.platform.commons.util.ReflectionUtils;
-import org.testcontainers.containers.PostgreSQLContainer;
+import org.testcontainers.containers.GenericContainer;
 
 @Slf4j
-public class FactCastExtension implements Extension, BeforeEachCallback {
+public class FactCastExtension
+    implements Extension,
```
[the remark this answers is not in the document]
Comment (uweschaefer) [reply factcast-test/src/main/java/org/factcast/test/FactCastExtension.java:30]: does not hurt and is more explicit, right?

File: factcast-test/src/main/java/org/factcast/test/FactCastExtension.java (hunk 1 above)
Comment (uweschaefer) [reply factcast-test/src/main/java/org/factcast/test/FactCastExtension.java:77]: needed due to potentially missing dependency to the redis modules

File: factcast-factus/src/main/java/org/factcast/factus/FactusImpl.java [hunk 4]
```diff
@@ -363,7 +363,9 @@ public void onNext(@NonNull Fact element) {
                 () -> {
                   handler.apply(element);
                   factId.set(element.id());
```
[the remark this answers is not in the document]
Comment (uweschaefer) [reply factcast-factus/src/main/java/org/factcast/factus/FactusImpl.java:365]: the scope and purpose of atomic refs is different to making a method synchronized or synchronizing on a block.
Even if it is not, if you can get away with Atomic* this is potentially faster due to less overhead. (mechanical sympathy)

File: factcast-factus/src/test/java/org/factcast/factus/projector/DefaultProjectorTest.java [hunk 5]
```diff
@@ -382,6 +374,7 @@ void apply(SimpleEvent e) {}
   }
 
   class NonStaticClass$MockitoMock extends NonStaticClass {
```
[the remark this answers is not in the document]
Comment (uweschaefer) [reply factcast-factus/src/test/java/org/factcast/factus/projector/DefaultProjectorTest.java:376]: we need to escalate from proxies to the classes they are wrapping. this is done by checking the classname for good hints on synthetics, like $MockitoMock or something CGLib-isch. this is what is tested here.

**Review by samba2 — APPROVED** [review]
lgtm

**Pull request merged**